File: examples/04-kafka/python/outbox_activity.py
import json

from temporalio import activity

# `db` is a transaction-capable handle (e.g. SQLAlchemy session / connection)
# injected into the Activity instance; `order_repository` / `outbox_repository`
# are repositories that enlist in the same transaction.


@activity.defn
async def write_order_and_outbox(order: dict) -> None:
    # One database transaction: business row and outbox row commit together.
    # A separate relay (or Debezium CDC) later ships the outbox row to Kafka, so
    # the DB write and the event publish can never diverge.
    with db.transaction():
        order_repository.save(order)
        outbox_repository.save(
            topic="order-events",
            key=order["id"],
            payload=json.dumps({"type": "OrderAccepted", "id": order["id"]}),
        )

File: examples/04-kafka/go/outbox_activity.go
package kafka

import (
	"context"
	"encoding/json"
)

// Order is the business row written by the Activity.
type Order struct {
	ID string
}

// WriteOrderAndOutbox writes the business row and an outbox row in ONE database
// transaction, so the row and its event can never diverge. A separate relay (or
// Debezium CDC) ships the outbox row to Kafka later — the Activity itself does
// no Kafka I/O, just an atomic local commit.
func WriteOrderAndOutbox(ctx context.Context, order Order) error {
	tx, err := db.BeginTx(ctx)
	if err != nil {
		return err
	}
	defer tx.Rollback()

	if err := orderRepository.Save(tx, order); err != nil {
		return err
	}
	payload, _ := json.Marshal(map[string]string{"type": "OrderAccepted", "id": order.ID})
	if err := outboxRepository.Save(tx, "order-events", order.ID, payload); err != nil {
		return err
	}
	return tx.Commit()
}
